tests/test_etl.py

```diff
@@ -0,0 +1,105 @@
+
+def fn(*args, **kwargs):
+    logger.debug(f"Finalizer called with {args} {kwargs}")
+    calls["fn"] += 1
+
+class TestETLException(Exception):
+    pass
+
+def error_fn(*args, **kwargs):
+    logger.debug(f"Finalizer called with {args} {kwargs}, raising error")
+    raise TestETLException("test")
+
+class TestETL(unittest.TestCase):
+    @unittest.skipIf(sys.version_info < (3, 6), "Only testing under Python 3.6+")
+    def test_etl(self):
+        from unittest.mock import MagicMock, patch
+        import dcplib.etl
+
+        files = [
+            {"name": hex(i), "content-type": "application/json", "uuid": hex(i), "version": "1", "sha256": "0"}
+            for i in range(256)
+        ]
+
+        class MockDSSClient:
+            host = "localhost"
+            swagger_url = "swagger_url"
+
+            class MockDSSMethod:
+                def __call__(self, es_query, replica):
+                    return {"total_hits": 1}
+
+                def iterate(self, es_query, replica, per_page):
+                    for i in range(4):
+                        yield {"bundle_fqid": f"a{i}.b"}
+
+            post_search = MockDSSMethod()
+
+            def get(self, url, params):
+                if "files" in url:
+                    payload = {}
+                else:
+                    payload = {"bundle": {"files": files}}
+                res = Response()
+                res.status_code = 200
+                res.raw = io.BytesIO(json.dumps(payload).encode())
+                return res
+
+        dcplib.etl.http = MockDSSClient()
+        with tempfile.TemporaryDirectory() as td:
+            e = dcplib.etl.DSSExtractor(staging_directory=td,
+                                        content_type_patterns=["application/json"],
+                                        filename_patterns=["*.json"],
+                                        dss_client=MockDSSClient())
+            e.extract(transformer=tf,
+                      loader=ld,
+                      finalizer=fn,
+                      query={"test": True},
+                      max_workers=2,
+                      max_dispatchers=1)
+            self.assertEqual(calls["tf"], 4)
+            self.assertEqual(calls["ld"], 4)
+            self.assertEqual(calls["fn"], 1)
+
+            e = dcplib.etl.DSSExtractor(staging_directory=td,
+                                        content_type_patterns=["application/json"],
+                                        filename_patterns=["*.json"],
+                                        dss_client=MockDSSClient())
+            e.extract(transformer=tf,
+                      loader=ld,
+                      finalizer=fn,
+                      query={"test": True},
+                      max_workers=2,
+                      max_dispatchers=1,
+                      dispatch_executor_class=concurrent.futures.ProcessPoolExecutor)
+            self.assertEqual(calls["tf"], 4)
+            self.assertEqual(calls["ld"], 4)
```

Comment:
`calls` doesn't get incremented if it's run in a separate process?

Reply:
Correct.